### ISSUE TITLE
Player.STATE_ENDED when prepare next track?

### ISSUE THREAD

**Issue by master255** [google/ExoPlayer#8192]
@ojw28 , I am trying to implement your new track switching system.
It doesn't behave logically.
1. I turn on the track:
MediaItem mediaItem = MediaItem.fromUri(Uri.parse(path));
                if (position > 0)
                    mMediaPlayer.setMediaItem(mediaItem, position);
                else
                    mMediaPlayer.setMediaItem(mediaItem, true);
                mMediaPlayer.prepare();
2. I am waiting: Player.STATE_READY  and do mMediaPlayer.clearMediaItems();
3. Then I turn on the next track at once:
MediaItem mediaItem = MediaItem.fromUri(Uri.parse(path));
                if (position > 0)
                    mMediaPlayer.setMediaItem(mediaItem, position);
                else
                    mMediaPlayer.setMediaItem(mediaItem, true);
                mMediaPlayer.prepare();
4. The Player.STATE_ENDED event is triggered. Why? The first track was not finished. The second track has not started playing yet.
5. The Player.STATE_READY event is not triggered. How do I know that the player has finished onPrepare?

I think the implementation of the new playlist system will be delayed...
There are many questions.
And I think most people will not use this functionality. Everyone needs to play one file. The playlist is a part of the application, not the playback module (ExoPlayer).

The problem does not appear if I use the old method:
mMediaPlayer.prepare(progressiveSource.createMediaSource(new MediaItem.Builder().setUri(Uri.parse(path)).build()), !haveStartPosition, false);

**Comment by marcbaechinger** [comment]
In step 2 the media items are removed with `clearMediaItems()` which makes the player having an empty playlist and the player transitions to `STATE_ENDED`.  If you do not call `clearMediaItems()`  but instead directly use `setMediaItems` you will not see `STATE_ENDED` but the player goes from `STATE_READY` to `STATE_BUFFERING` directly.

**Comment by master255** [comment]
@marcbaechinger Ok. I want to play one track. Then quickly the second one. What do I need to do?
When I start the first track, I don't know what the second track will be. I can't use the playlist.

**Comment by master255** [comment]
@marcbaechinger Another bug. If I do not call:
mMediaPlayer.clearMediaItems();
then even if I call a new track with a position reset, the position is not reset but resumed:
mMediaPlayer.setMediaItem(mediaItem, true) - position reset does not work.
Even if I call:
mMediaPlayer.setMediaItem(mediaItem);
the position is not reset.

**Comment by master255** [comment]
@marcbaechinger I see you have forgotten about this variable.
![1](https://user-images.githubusercontent.com/5380115/98587733-9bae3680-22db-11eb-9170-bcad1450b754.PNG)

**Comment by marcbaechinger** [comment]
Yes you are right, there was a bug regarding the `resetPosition` not being honored (#8024). This has been fixed and released with [2.12.1.](https://github.com/google/ExoPlayer/blob/release-v2/RELEASENOTES.md#2121-2020-10-23) So in case you are using 2.12.0 an upgrade should help. Can you check what version you are using?

> I want to play one track. Then quickly the second one. What do I need to do?

I just tested your case with the actual code (2.12.1) and it worked fine. I do an initial preparation with one or more media items. Then when the user presses a button or similar I just do:

```
player.setMediaItem(MediaItem.fromUri("https://html5demos.com/assets/dizzy.mp4"), /* resetPosition= */ true);
``` 

This replaces all items in the current playlist with the media item above, prepares the new source, resets the position (starts from position 0 with the new stream) and continues playback from there.

**Comment by master255** [comment]
@marcbaechinger It is difficult for me to compile all the codecs every time.
I currently have version 2.12.0. Ok. I'll check it out later.
I'll wait for version 2.12.2.
